Add brightness-mode-explainer.md.

rakuco opened the pull request:
Related to #129 and w3c/devicesensors-wg#51 for more context. The contents
come from
https://docs.google.com/document/d/1skbEBafMjKnYVC7UoJfRgnnhAkp-jx0k_Z4j6eZeOwc/edit?usp=sharing
which was used while hashing out the basics and the structure.

Now that the contents have matured a bit, it is time to move this to an
explainer and continue iterating here.

Commit:
Add brightness-mode-explainer.md.

Related to #129 and w3c/devicesensors-wg#51 for more context. The contents
come from
https://docs.google.com/document/d/1skbEBafMjKnYVC7UoJfRgnnhAkp-jx0k_Z4j6eZeOwc/edit?usp=sharing
which was used while hashing out the basics and the structure.

Now that the contents have matured a bit, it is time to move this to an
explainer and continue iterating here.

rakuco commented:
+@willmorgan who doesn't seem to be part of the WG so I couldn't formally add as a reviewer.

Compared to a regular explainer, this one has a slightly different structure: we don't have the API fully hashed out, and the "considered alternatives" section was renamed to "proposed solutions", which serves a somewhat different purpose and indicates that we are still open to the ideas presented there.

anssiko reviewed (approved):
Thanks! I like there are multiple proposed solutions. It really shows this seemingly simple API has received careful scrutiny from various perspectives.

This is how most specs are made, it is just that usually the sausage making process is not this carefully documented :)

marcoscaceres commented on brightness-mode-explainer.md:
````diff
@@ -0,0 +1,129 @@
+  - [\<video\> integration to allow e.g. granular brightness control](https://github.com/w3c/screen-wake-lock/issues/129#issuecomment-926603108) is an instance of the above. This is most likely better handled elsewhere and should probably be a UA-specific control.
+
+## API
+### Web IDL
+
+Extend the [Screen interface](https://drafts.csswg.org/cssom-view/#the-screen-interface) with a new operation, inspired by the [Wake Lock API](https://developer.mozilla.org/en-US/docs/Web/API/WakeLockSentinel). Something like:
+
+```webidl
+[SecureContext]
+partial interface Screen {
+  Promise<ScreenBrightnessSentinel> requestBrightnessIncrease();
+};
+
+interface ScreenBrightnessSentinel {
+  Promise<undefined> release();
+};
+```
+
````
I respectfully object to this being added here as we never settled on an API (i.e., we wouldn't need this document if we had settled on a particular API already). 

I asked that this be moved to the "Proposed Solutions" section, along with the other proposed solutions. 

This should be moved to the the proposed solutions section.

rakuco replied:
Done. I honestly thought we'd settled on this being the current proposal (which is why I'd originally listed the other ideas as "considered alternatives" rather than "proposed solutions" in the document), but this makes it easier to understand why we have "proposed solutions" in the first place.

marcoscaceres reviewed (changes requested):
The "API" is just another proposed solution.

marcoscaceres commented:
(despite my objection, I'm still really thankful that this is coming along ... great work @rakuco and team)

anssiko commented:
Thanks @marcoscaceres, I think it is reasonable to put all the ideas under the proposed solutions, including the API draft.

@rakuco was simply implementing the group's resolution https://www.w3.org/2021/11/17-dap-minutes.html#r01 that did not perhaps have the best wording ever. That error is on me, my apologies for that.

Commit:
Make the Screen-based idea a proposed solution as well.

willmorgan reviewed (approved):
Thanks for amalgamating the various voices here @rakuco.

beaufortfrancois commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,128 @@
+};
+
+interface ScreenBrightnessSentinel {
+  Promise<undefined> release();
```
Nit: I'm not sure returning a promise is required.

willmorgan replied:
It's useful to do this and have the promise fulfil when the release completes. This would be useful if vendors implement brightness changes as transitions rather than a binary toggle.

Entering and exiting fullscreen is asynchronous so this would work in a similar way.

beaufortfrancois replied:
From what I can see, native APIs used to change the device brightness are not asynchronous.

beaufortfrancois replied:
See Android for instance: https://developer.android.com/reference/android/view/WindowManager.LayoutParams#screenBrightness

willmorgan replied:
@beaufortfrancois You're 100% correct. I don't know whether this is going to change in the future and therefore if it's worth having the vendor return a fulfilled promise in case it does?

rakuco replied:
This sounds similar to `WakeLockSentinel.release()`, which currently returns a promise even though it is always ends up being fulfilled synchronously. Returning a promise does make room for more changes in the future, but I agree it looks weird if we don't do anything asynchronous in the method.

marcoscaceres replied:
From experience with screen orientation lock, I'd also be inclined to make it async. 

Although changing the brightness is sync at the OS level, there might be some IPC involved and stuff can get out of sync. Worst case, we just return `Promise.resolve()`.

rakuco replied:
Let's keep this as-is for now; there is still time to change it before anyone implements any of the ideas here, and using a Promise is more future-proof.

beaufortfrancois commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,128 @@
+  try {
+    const sentinel = await screen.requestBrightnessIncrease();
+    console.log("The display's brightness level has increased");
+    await sentinel.release();
```
```suggestion
    // …and release it after 5s.
    window.setTimeout(() => {
      await sentinel.release();
    }, 5000);

rakuco replied:
Done, thanks!

beaufortfrancois reviewed (approved):
LGTM with nits

kenchris commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,128 @@
+
+- Things to consider in the spec proposal
+  - UAs should be free to limit maximum brightness level and/or do other things like increase contrast.
+  - Idea is to focus on mobile first (i.e. not worry about external displays). How to indicate that in the spec? Does it *need* to be indicated in the spec?
```
It might make sense to file a TAG early review request, highlighting that you would like feedback on these points.

kenchris commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,128 @@
+  - Should this trigger "`prefers-contrast: more`" in [CSS Media Queries](https://drafts.csswg.org/mediaqueries-5/#prefers-contrast)?
+- From [https://github.com/w3c/devicesensors-wg/issues/51](https://github.com/w3c/devicesensors-wg/issues/51)
+  - Brightness levels
+    - How bright is too bright? Consider 100% brightness with HDR displays, for example.
```
I guess you could talk about specific nit values

rakuco replied:
What do you mean by "nit values"?

kenchris replied:
It is a unit for light intensity: https://en.wikipedia.org/wiki/Candela_per_square_metre - it is very common to use it for displays.

Good article: https://www.maketecheasier.com/what-is-nit-of-screen-brightness/

>"Technically, a device starts counting as “sunlight-readable” when it hits at least 1,000 nits, but very few mobile displays go that high. As a general rule, anything above 400 to 500 nits will do pretty well on a sunny day, but at 200 nits, you may have to find some shade to answer texts."

rakuco replied:
I see. I'm used to "nit' being short for "nitpick", which makes some of those sentences very funny to read :-)

Do you know if nits are exposed by any platform?

kenchris replied:
I don't know sorry. But on Android you at least seem to be able to get the max range: https://stackoverflow.com/questions/4544967/get-preferred-screen-brightness-in-android/8114307#8114307

kenchris commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,128 @@
+
+- Provide the ability to request an increase to the display's brightness. This could be a UA-controlled maximum (which *could* be the maximum display brightness), either indefinitely or for a period of time.
+- Provide the ability to release the request so that the device's brightness returns to its pre-request value (i.e. hand back control to OS).
+- Handle error cases, where such requests are denied or not possible.
```
low battery?

rakuco replied:
I've added it to this sentence while also clarifying that we need to be careful not to expose a low battery state to avoid fingerprinting.

Commit:
Add timeout to the code example

Co-authored-by: François Beaufort <beaufort.francois@gmail.com>

Commit:
Mention low battery in the Goals section.

marcoscaceres commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,131 @@
+# Design Document: the need to control screen brightness
+
+_As discussed in the [DAS WG 2021-11-17](https://www.w3.org/2021/11/17-dap-minutes.html) meeting about Screen Brightness and the Screen Wake Lock API, the idea of this document is to make it easier for the meeting attendees to work on an explainer than if this had been committed to GitHub directly. Once this is done the goal is to move the contents to a proper explainer on GitHub._
+
+Previously:
+- https://github.com/WICG/proposals/issues/17
+- https://github.com/w3c/screen-wake-lock/issues/129
+- https://github.com/w3c/devicesensors-wg/issues/51
+- https://www.w3.org/2021/11/17-dap-minutes.html
```
We should probably move all this to the end of this document and move the Introduction's text here . This stuff is  only relevant for us WG members, but not for anyone else.

rakuco replied:
Done. I've dropped the italicized paragraph since we're already on GitHub and moved the links to a new section at the end of the document.

marcoscaceres commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,131 @@
+There is currently a need for web applications to increase the brightness of a device's screen.
+
+This would help address the following use cases:
+- a user scans QR/barcode under various lighting conditions.
```
```suggestion
- a user scans QR/barcode under various lighting conditions, in which case boosting the screen brightness creates more contrast.
```

marcoscaceres commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,131 @@
+
+This would help address the following use cases:
+- a user scans QR/barcode under various lighting conditions.
+- remote medical, where increasing the screen brightness could assist in remote examination
```
This is a stretch (at least, none of us are medical doctors here)... I suggest we remove this. I also don't think any developer actually proposed this? 

```suggestion
```

rakuco replied:
This came from WICG/proposals#17 so I'd like to confirm with @willmorgan before dropping this line.

willmorgan replied:
I am not a medical doctor or healthtech developer, but I can think of various body parts or mobility restrictions where front torch or maximum screen brightness would be useful. Not super compelling though.

rakuco replied:
Done, I've removed this item from the list.

marcoscaceres commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,131 @@
+This would help address the following use cases:
+- a user scans QR/barcode under various lighting conditions.
+- remote medical, where increasing the screen brightness could assist in remote examination
+- biometric security apps, where increasing the screen brightness can help illuminate features to get better imagery from the front facing camera
```
```suggestion
- Increasing the screen brightness can help illuminate a user's facial features, improving the viability of certain application that rely on the the front facing camera (e.g., [make up mirror](https://play.google.com/store/apps/details?id=mmapps.mirror.pro&hl=en&gl=US) style apps, and applications that can scan people's facial features to perform some particular action)
```

rakuco replied:
Done.

marcoscaceres commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,131 @@
+- a user scans QR/barcode under various lighting conditions.
+- remote medical, where increasing the screen brightness could assist in remote examination
+- biometric security apps, where increasing the screen brightness can help illuminate features to get better imagery from the front facing camera
+- increasing contrast for readability for the visually impaired
```
I think this has a general benefit, not just for "the visually impaired"

However, phones already dynamically adjust screen brightness based on lighting conditions... so I think this is not a good use case. 

I suggest we drop this as it's a foot gun. 

```suggestion
```

The counter would be, it's making it **too bright** for sensitive eyes. 

Consider, I want to read an article at night and not disturb people around me by blasting up the brightness.

rakuco replied:
This also came from WICG/proposals#17 so I'd also like to double-check with @willmorgan

reillyeon replied:
I would drop this because it should be user agent and operating system accessibility features doing this adjustment rather than individual sites. The other use cases focus on scenarios where the site has specific knowledge that greater brightness is required while this one does not.

willmorgan replied:
I agree that this should be removed. I get the sense that there are more appropriate ways of facilitating VI users.

rakuco replied:
Done, item removed.

marcoscaceres commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,131 @@
+- remote medical, where increasing the screen brightness could assist in remote examination
+- biometric security apps, where increasing the screen brightness can help illuminate features to get better imagery from the front facing camera
+- increasing contrast for readability for the visually impaired
+- [make up mirror](https://play.google.com/store/apps/details?id=mmapps.mirror.pro&hl=en&gl=US) style apps
```
Combined with above... 

```suggestion
```

rakuco replied:
Done.

marcoscaceres commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,131 @@
+- increasing contrast for readability for the visually impaired
+- [make up mirror](https://play.google.com/store/apps/details?id=mmapps.mirror.pro&hl=en&gl=US) style apps
+
+[Screen Wake Lock API issue 129](https://github.com/w3c/screen-wake-lock/issues/129) was filed back in 2018, and since then there has been feedback from developers working on multiple areas about the usefulness of an API that addresses this use case.
```
Move this up...

```suggestion
```

marcoscaceres commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,131 @@
+
+## Introduction
+
+There is currently a need for web applications to increase the brightness of a device's screen.
```
```suggestion
Since 2018, the working group has received [significant requests](https://github.com/w3c/screen-wake-lock/issues/129) from web developers surrounding the need to increase the brightness of a device's screen.
```

marcoscaceres commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,131 @@
+
+There is currently a need for web applications to increase the brightness of a device's screen.
+
+This would help address the following use cases:
```
```suggestion
For developers, it would help address the following use cases:

```

marcoscaceres commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,131 @@
+
+## Open design issues
+
+- Things to consider in the spec proposal
```
```suggestion
The following issues remain open for discussion:
```

marcoscaceres commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,131 @@
+## Open design issues
+
+- Things to consider in the spec proposal
+  - UAs should be free to limit maximum brightness level and/or do other things like increase contrast.
```
```suggestion
  - should UAs be free to limit maximum brightness level and/or do other things like increase contrast?
```

marcoscaceres commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,131 @@
+
+- Things to consider in the spec proposal
+  - UAs should be free to limit maximum brightness level and/or do other things like increase contrast.
+  - Idea is to focus on mobile first (i.e. not worry about external displays). How to indicate that in the spec? Does it *need* to be indicated in the spec?
```
```suggestion
  - Should this be mobile first/only? (i.e. not worry about external displays).
```

marcoscaceres commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,131 @@
+- From [https://github.com/w3c/devicesensors-wg/issues/51](https://github.com/w3c/devicesensors-wg/issues/51)
+  - Brightness levels
+    - How bright is too bright? Consider 100% brightness with HDR displays, for example.
+    - Take a granular value or an enum?
```
```suggestion
    - Take a discrete or continuous value?
```

marcoscaceres commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,131 @@
+    - Take a granular value or an enum?
+    - Related to whether script should be allowed to reduce brightness.
+  - Permission model
+    - Probably require user gesture (request but not consume it).
```
```suggestion
    - Would it require user gesture (request but not consume it)?
```

marcoscaceres commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,131 @@
+    - Related to whether script should be allowed to reduce brightness.
+  - Permission model
+    - Probably require user gesture (request but not consume it).
+    - How to deal with multiple or external displays
```
This is already above.

```suggestion
```

marcoscaceres commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,131 @@
+- [make up mirror](https://play.google.com/store/apps/details?id=mmapps.mirror.pro&hl=en&gl=US) style apps
+
+[Screen Wake Lock API issue 129](https://github.com/w3c/screen-wake-lock/issues/129) was filed back in 2018, and since then there has been feedback from developers working on multiple areas about the usefulness of an API that addresses this use case.
+
```
We might want to make a case here that native applications support this, and maybe provide a few examples (e.g., I think Eventbright does this for QR codes, but need to check).

rakuco replied:
That's a very good idea. @willmorgan @kenchris @beaufortfrancois help here would be appreciated :-)

willmorgan replied:
Starbucks, Tesco App (for clubcard), [UK Home Office Document Check App](https://play.google.com/store/apps/details?id=uk.gov.HomeOffice.ho1&hl=en_GB&gl=US), etc.

rakuco replied:
I've added a subitem with a few links to the Google Play store for the Starbucks app, the UK Home Office Document Check App and the Dutch government's CoronaCheck app.

I don't have first-hand experience with the Starbucks app so I hope I've linked to the right app and that it actually does what we think it does :-) I believe Will when it comes to the Document Check App, and I've used the Dutch CoronaCheck app myself and know it changes the device's screen brightness level. If anyone has links to other apps (like the Tesco one), please let me know.

marcoscaceres commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,131 @@
+
+- Provide the ability to request an increase to the display's brightness. This could be a UA-controlled maximum (which *could* be the maximum display brightness), either indefinitely or for a period of time.
+- Provide the ability to release the request so that the device's brightness returns to its pre-request value (i.e. hand back control to OS).
+- Handle error or low battery cases, where such requests are denied or not possible. The API must not make it possible to determine whether a request failed due to low battery to avoid fingerprinting.
```
This is not really a fingerprinting concern... more of a privacy concern. 

```suggestion
- Handle error or low battery cases, where such requests are denied or not possible. For privacy, the API must not make it possible to determine whether a request failed due to low battery.
```

marcoscaceres commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,131 @@
+
+## Proposed Solutions
+
+This was initially discussed in [Screen Wake Lock API issue 129](https://github.com/w3c/screen-wake-lock/issues/129) back in 2018 and also proposed as a [separate API in WICG](https://github.com/WICG/proposals/issues/17). This topic was also discussed at multiple TPACs under the Devices and Sensors WG. As such, several different API shapes have been considered over the years.
+
+At this point, none have been fully discarded, and we are open to feedback about adopting some of the ideas described here.
```
```suggestion
The following represent some rough proposals that could address the use cases using various web technologies.  

We present them here only to foster discussion, and the working group has not settled on any particular one. We are open to feedback to pursuing any of the proposals below. Or, if there is a better alternative solution we have not considered, we would be open to hearing it!
```

marcoscaceres commented on brightness-mode-explainer.md:
````diff
@@ -0,0 +1,131 @@
+
+### Screen Wake Lock integration
+
+This was considered for a long time until the [2021-11-17 DAS WG meeting to discuss the topic](https://www.w3.org/events/meetings/0f623aa1-2026-4366-846b-c2faedda4180). The idea was to extend the existing Screen Wake Lock API and tie the change in brightness to a WakeLockSentinel. Something like:
+
+``` javascript
+const lock = await navigator.wakeLock.request({ increaseBrightness: true });
+```
````
```suggestion
The idea was to extend the existing Screen Wake Lock API and tie the change in brightness to a WakeLockSentinel. Something like:

``` javascript
const lock = await navigator.wakeLock.request({ increaseBrightness: true });
```

This was considered for a long time until the [2021-11-17 DAS WG meeting to discuss the topic](https://www.w3.org/events/meetings/0f623aa1-2026-4366-846b-c2faedda4180).
```

marcoscaceres commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,131 @@
+
+### navigator.screenBrightness
+
+Originally proposed in [WICG issue 17](https://github.com/WICG/proposals/issues/17). The API proposed there had a larger surface that allowed reading the brightness value (leading to potential fingerprinting issues).
+
```
```suggestion
The API proposed there had a larger surface that allowed reading the brightness value (leading to potential fingerprinting issues). Originally proposed in [WICG issue 17](https://github.com/WICG/proposals/issues/17).

```

marcoscaceres commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,131 @@
+
+### requestFullscreen() integration
+
+A mixture of [whatwg/fullscreen](https://fullscreen.spec.whatwg.org/) issues [185](https://github.com/whatwg/fullscreen/issues/185) and [198](https://github.com/whatwg/fullscreen/issues/198). The idea would be to do something like
+
```
```suggestion
The idea to integrate with `.requestFullscreen()`, to do something like:

```

marcoscaceres commented on brightness-mode-explainer.md:
````diff
@@ -0,0 +1,131 @@
+```
+
+To take advantage of existing UA privacy mitigations and UX indications that would show to the user that a web application is increasing the brightness, and leaving full screen mode would make it clear that the UA should stop increasing the device's screen's brightness level.
+
````
```suggestion
See [whatwg/fullscreen](https://fullscreen.spec.whatwg.org/) issues [185](https://github.com/whatwg/fullscreen/issues/185) and [198](https://github.com/whatwg/fullscreen/issues/198).
```

marcoscaceres commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,131 @@
+
+Add something to getUserMedia() to bundle the request for brightness into the media capture request itself.
+
+The idea was dropped because this feature is not doing media capture and the need to ask for camera permission before e.g. entering full screen mode [complicates some use cases](https://github.com/w3c/screen-wake-lock/issues/129#issuecomment-858790397).
+
```
If the idea was dropped, it wouldn't be in this doc ;) Let's say: 


```suggestion
This is complicated because feature is not doing media capture, and the need to ask for camera permission before e.g. entering full screen mode [complicates some use cases](https://github.com/w3c/screen-wake-lock/issues/129#issuecomment-858790397).

```

rakuco replied:
Done.

Commit:
Incorporate review suggestion

Expand bullet point scanning and increased contrast.

Co-authored-by: Marcos Cáceres <marcos@marcosc.com>

Commit:
Incorporate review suggestions.

Merge make-up mirror use case with biometrical one and reword the whole
item.

Commit:
Update brightness-mode-explainer.md

Co-authored-by: Marcos Cáceres <marcos@marcosc.com>

Commit:
Update brightness-mode-explainer.md

Co-authored-by: Marcos Cáceres <marcos@marcosc.com>

Commit:
Update brightness-mode-explainer.md

Co-authored-by: Marcos Cáceres <marcos@marcosc.com>

Commit:
Update brightness-mode-explainer.md

Co-authored-by: Marcos Cáceres <marcos@marcosc.com>

Commit:
Update brightness-mode-explainer.md

Co-authored-by: Marcos Cáceres <marcos@marcosc.com>

Commit:
Update brightness-mode-explainer.md

Co-authored-by: Marcos Cáceres <marcos@marcosc.com>

Commit:
Incorporate review suggestion.

Reword the getUserMedia() paragraph following Marcos' suggestion.

Commit:
Apply suggestions from code review

Move whatwg/fullscreen links to the end of the section.

Co-authored-by: Marcos Cáceres <marcos@marcosc.com>

Commit:
Update brightness-mode-explainer.md

Co-authored-by: Marcos Cáceres <marcos@marcosc.com>

Commit:
Update brightness-mode-explainer.md

Co-authored-by: Marcos Cáceres <marcos@marcosc.com>

Commit:
Adjust previous commit.

GitHub's UI botched the suggestion and erased a paragraph that should not
have been removed.

Commit:
Apply suggestions from code review

Move link to Screen Wake Lock API issue 129 to the introduction.

Co-authored-by: Marcos Cáceres <marcos@marcosc.com>

Commit:
Apply suggestions from code review

Co-authored-by: Marcos Cáceres <marcos@marcosc.com>

Commit:
Add a missing empty line between headers.

Commit:
Move links from the beginning to the end of the text.

Drop paragraph explaining what the document is about; we have already moved
to GitHub.

marcoscaceres reviewed (approved):
Pre-approving modulo the changes I requested (pending feedback from @willmorgan). 

Looking really good team! 💪❤️

reillyeon commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,129 @@
+
+Some form of "scannable element" property. When an element with said property is visible, the UA would take care of increasing the brightness level.
+
+*Note: this is the least hashed out proposal at this point. It is unclear how this would work with zoom, permissions, low-battery mode, what happens when an element scrolls in/out of view, or even how to mitigate something like*
```
If we implemented one of the proposals above featuring a `requestIncreasedBrighness()` method then I would polyfill this proposal roughly as,

```
let sentinel;
element.onvisible = () => {
  sentinel = requestFullBrightness();
};
element.onhidden = () => {
  sentinel.release();
};
```

Polyfilled in this way all the same abuse mitigations would apply with the advantage that the browser has some insight into how the page content relates to the developer's desire for increased brightness and can do things like offer to increase brightness when the element is visible (instead of automatically doing so) or increase brightness gradually as the element enters the viewport.

rakuco replied:
Wouldn't that bypass the user gesture requirement and allow one to add the max-brightness style to `<body>`?

reillyeon replied:
Not necessarily. The browser could still require the user interact with the page in some way before the property took effect or else it would display its own UI prompting the user with something like "tap here to make this page easier to scan".

reillyeon reviewed (approved):
Approved with a few comments. Thank you everyone for your work on this issue.

anssiko commented:
Thanks everyone for your feedback. This doc received more scrutiny than I personally expected. I think that is a signal this feature is both important to people and that the integration into the platform is not trivial.

Now that the doc has taken a shape (@rakuco to merge), we want to seek review from:

- CSS WG for CSS bits per https://www.w3.org/2021/11/17-dap-minutes.html#r01

Optional:
- TAG for early review of open design issues
- An early privacy review with PING
- Other horizontal groups on a need to know basis

willmorgan commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,129 @@
+To take advantage of existing UA privacy mitigations and UX indications that would show to the user that a web application is increasing the brightness, and leaving full screen mode would make it clear that the UA should stop increasing the device's screen's brightness level.
+See [whatwg/fullscreen](https://fullscreen.spec.whatwg.org/) issues [185](https://github.com/whatwg/fullscreen/issues/185) and [198](https://github.com/whatwg/fullscreen/issues/198).
+
+### getUserMedia() integration
```
I recall this being put aside for a number of reasons as there's no MediaStream you can get out of screen brightness, and it'd be unclear how to deal with constraint application / failures etc.

I'm not particularly passionate about removing it (or including it) either way, though.

rakuco replied:
I've expanded the text below and mentioned the MediaStream bit from your comment. Please let me know if you feel there's more to add.

Commit:
Remove a few use cases from the least.

There was some agreement that these are not very good ones.

Commit:
Add examples of mobile apps that change the screen's brightness level.

Commit:
Elaborate a bit more on why getUserMedia() is not a good option.

rakuco commented:
Thanks everyone, I also appreciate all the attention that's been given to this document! I've pushed 3 new commits today addressing the open items we had:
- removed a few use cases that we did not agree upon
- added a bit more text to the getUserMedia() proposal
- added examples of mobile apps that change the device's screen brightness level

Reilly's point about the CSS property and how it could be polyfilled remains open, but I don't see it as a blocker. I'll merge this PR tomorrow to give everyone time to take one last look.

marcoscaceres commented:
About additional reviews, once merged it would be good to share the document in Apple’s and Mozilla’s standards position repos to see if it might be of interest to them. They are generally good at identifying things we may have overlooked or possibly provide novel solutions that leverage existing standards.

tomayac commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,128 @@
+
+For developers, it would help address the following use cases:
+
+- a user scans QR/barcode under various lighting conditions, in which case boosting the screen brightness creates more contrast.
```
```suggestion
- a user scans a QR/barcode under various lighting conditions, in which case boosting the screen brightness creates more contrast.
```

tomayac commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,128 @@
+
+- a user scans QR/barcode under various lighting conditions, in which case boosting the screen brightness creates more contrast.
+  - this is already possible on native mobile apps such as [Starbucks'](https://play.google.com/store/apps/details?id=com.starbucks.mobilecard), [the UK Home Office's Document Check app](https://play.google.com/store/apps/details?id=uk.gov.HomeOffice.ho1) and the [Dutch government's CoronaCheck app](https://play.google.com/store/apps/details?id=nl.rijksoverheid.ctr.holder).
+- increasing the screen brightness can help illuminate a user's facial features, improving the viability of certain application that rely on the the front facing camera (e.g., [make up mirror](https://play.google.com/store/apps/details?id=mmapps.mirror.pro&hl=en&gl=US) style apps, and applications that can scan people's facial features to perform some particular action)
```
```suggestion
- increasing the screen brightness can help illuminate a user's facial features, improving the viability of certain application that rely on the front facing camera (e.g., [make up mirror](https://play.google.com/store/apps/details?id=mmapps.mirror.pro&hl=en&gl=US) style apps, and applications that can scan people's facial features to perform some particular action).
```

tomayac commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,128 @@
+## Open design issues
+
+The following issues remain open for discussion:
+  - should UAs be free to limit maximum brightness level and/or do other things like increase contrast?
```
```suggestion
  - should UAs be free to limit the maximum brightness level and/or do other things like increase the contrast?
```

tomayac commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,128 @@
+
+The following issues remain open for discussion:
+  - should UAs be free to limit maximum brightness level and/or do other things like increase contrast?
+  - Should this be mobile first/only? (i.e. not worry about external displays).
```
```suggestion
  - Should this be mobile first/only? (i.e., not worry about external displays)?
```

tomayac commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,128 @@
+The following issues remain open for discussion:
+  - should UAs be free to limit maximum brightness level and/or do other things like increase contrast?
+  - Should this be mobile first/only? (i.e. not worry about external displays).
+  - Should this trigger "`prefers-contrast: more`" in [CSS Media Queries](https://drafts.csswg.org/mediaqueries-5/#prefers-contrast)?
```
```suggestion
  - Should this trigger `prefers-contrast: more` in [CSS Media Queries](https://drafts.csswg.org/mediaqueries-5/#prefers-contrast)?
```

tomayac commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,128 @@
+    - Take a discrete or continuous value?
+    - Related to whether script should be allowed to reduce brightness.
+  - Permission model
+    - Would it require user gesture (request but not consume it)?
```
```suggestion
    - Would it require a user gesture (request but not consume it)?
```

tomayac commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,128 @@
+    - Would it require user gesture (request but not consume it)?
+  - While brightness changes
+    - What if users change the brightness level while the lock is held?
+  - When dropping screen brightness request
```
```suggestion
  - When dropping a screen brightness request
```

tomayac commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,128 @@
+## Goals
+
+- Provide the ability to request an increase to the display's brightness. This could be a UA-controlled maximum (which *could* be the maximum display brightness), either indefinitely or for a period of time.
+- Provide the ability to release the request so that the device's brightness returns to its pre-request value (i.e. hand back control to OS).
```
```suggestion
- Provide the ability to release the request so that the device's brightness returns to its pre-request value (i.e., hand back control to OS).
```

tomayac commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,128 @@
+
+## Non-goals
+
+- Reading screen brightness level.
```
```suggestion
- Reading the current screen brightness level.
```

tomayac commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,128 @@
+## Non-goals
+
+- Reading screen brightness level.
+- Adjusting display brightness level to arbitrary values (absolute or relative).
```
```suggestion
- Adjusting the display brightness level to arbitrary values (absolute or relative).
```

tomayac commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,128 @@
+
+- Reading screen brightness level.
+- Adjusting display brightness level to arbitrary values (absolute or relative).
+  - [\<video\> integration to allow e.g. granular brightness control](https://github.com/w3c/screen-wake-lock/issues/129#issuecomment-926603108) is an instance of the above. This is most likely better handled elsewhere and should probably be a UA-specific control.
```
```suggestion
  - [\<video\> integration to allow, e.g., granular brightness control](https://github.com/w3c/screen-wake-lock/issues/129#issuecomment-926603108) is an instance of the above. This is most likely better handled elsewhere and should probably be a UA-specific control.
```

tomayac commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,128 @@
+
+## Proposed Solutions
+
+The following represent some rough proposals that could address the use cases using various web technologies.  
```
```suggestion
The following represents some rough proposals that could address the use cases using various web technologies.  
```

tomayac commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,128 @@
+
+The following represent some rough proposals that could address the use cases using various web technologies.  
+
+We present them here only to foster discussion, and the working group has not settled on any particular one. We are open to feedback to pursuing any of the proposals below. Or, if there is a better alternative solution we have not considered, we would be open to hearing it!
```
```suggestion
We present them here only to foster discussion, and the working group has not settled on any particular one. We are open to feedback to pursue any of the proposals below. Or, if there is a better alternative solution we have not considered, we would be open to hearing it!
```

tomayac commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,128 @@
+
+### Screen IDL interface extension
+
+Extend the [Screen interface](https://drafts.csswg.org/cssom-view/#the-screen-interface) with a new operation, inspired by the [Wake Lock API](https://developer.mozilla.org/en-US/docs/Web/API/WakeLockSentinel). Something like:
```
```suggestion
Extend the [`Screen` interface](https://drafts.csswg.org/cssom-view/#the-screen-interface) with a new operation, inspired by the [Wake Lock API](https://developer.mozilla.org/en-US/docs/Web/API/WakeLockSentinel). Something like:
```

tomayac commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,128 @@
+
+### Screen Wake Lock integration
+
+The idea was to extend the existing Screen Wake Lock API and tie the change in brightness to a WakeLockSentinel. Something like:
```
```suggestion
The idea was to extend the existing Screen Wake Lock API and tie the change in brightness to a `WakeLockSentinel`. Something like:
```

tomayac commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,128 @@
+
+This was considered for a long time until the [2021-11-17 DAS WG meeting to discuss the topic](https://www.w3.org/events/meetings/0f623aa1-2026-4366-846b-c2faedda4180).
+
+### navigator.screenBrightness
```
```suggestion
### `navigator.screenBrightness`
```

tomayac commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,128 @@
+
+The API proposed there had a larger surface that allowed reading the brightness value (leading to potential fingerprinting issues). Originally proposed in [WICG issue 17](https://github.com/WICG/proposals/issues/17).
+
+### requestFullscreen() integration
```
```suggestion
### `requestFullscreen()` integration
```

tomayac commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,128 @@
+
+### requestFullscreen() integration
+
+The idea to integrate with `.requestFullscreen()`, to do something like:
```
```suggestion
The idea to integrate with `Element.requestFullscreen()`, to do something like:
```

tomayac commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,128 @@
+
+### getUserMedia() integration
+
+Add something to getUserMedia() to bundle the request for brightness into the media capture request itself.
```
```suggestion
Add something to `getUserMedia()` to bundle the request for brightness into the media capture request itself.
```

tomayac commented on brightness-mode-explainer.md:
```diff
@@ -0,0 +1,128 @@
+
+Add something to getUserMedia() to bundle the request for brightness into the media capture request itself.
+
+This is complicated because this feature is not doing media capture (and there is no MediaStream to get out of screen brightness), and the need to ask for camera permission before e.g. entering full screen mode [complicates some use cases](https://github.com/w3c/screen-wake-lock/issues/129#issuecomment-858790397).
```
```suggestion
This is complicated because this feature is not doing media capture (and there is no `MediaStream` to get out of screen brightness), and the need to ask for camera permission before, e.g., entering full screen mode [complicates some use cases](https://github.com/w3c/screen-wake-lock/issues/129#issuecomment-858790397).
```

tomayac committed:
Apply suggestions from code review

rakuco commented:
Thanks for the edits, @tomayac! It's finally time to merge this PR \o/

beaufortfrancois commented:
> About additional reviews, once merged it would be good to share the document in Apple’s and Mozilla’s standards position repos to see if it might be of interest to them. They are generally good at identifying things we may have overlooked or possibly provide novel solutions that leverage existing standards.

@rakuco Do you want to share https://github.com/w3c/screen-wake-lock/blob/gh-pages/brightness-mode-explainer.md with Apple and Mozilla?

rakuco commented:
> @rakuco Do you want to share https://github.com/w3c/screen-wake-lock/blob/gh-pages/brightness-mode-explainer.md with Apple and Mozilla?

My plate's a bit full at the moment, so I'd appreciate it if someone like @marcoscaceres could take this up after having provided so much valuable feedback :-)

beaufortfrancois commented:
@marcoscaceres (gentle ping)

beaufortfrancois commented:
If @marcoscaceres is okay with that, I can start sending an email to webkit-dev@ and request a position at https://github.com/mozilla/standards-positions/

rakuco commented:
I'd say go for it :-)

beaufortfrancois commented:
Here it is:
- https://lists.webkit.org/pipermail/webkit-dev/2022-March/032160.html
- https://github.com/mozilla/standards-positions/issues/623

Feel free to star/subscribe to those threads to follow along.

rakuco commented:
Awesome, thanks!

beaufortfrancois commented:
FYI Mozilla folks have started giving feedback on this document at https://github.com/mozilla/standards-positions/issues/623
Feel free to join the conversation

anssiko commented:
I opened https://github.com/w3c/screen-wake-lock/issues/335 for API design discussion based on Mozilla's feedback.
